network/p2p/topic_validator_test.go

```diff
@@ -109,31 +107,349 @@ func TestTopicValidator(t *testing.T) {
+
+	authorizedSenderValidator := validator.AuthorizedSenderValidator(logger, func(pid peer.ID) (*flow.Identity, bool) {
+		fid, err := translator.GetFlowID(pid)
+		if err != nil {
+			return &flow.Identity{}, false
+		}
+		return ids.ByNodeID(fid)
+	})
+
+	// node1 is connected to node2, and the an1 is connected to node1
+	// an1 <-> sn1 <-> sn2
+	require.NoError(t, sn1.AddPeer(context.TODO(), *host.InfoFromHost(sn2.host)))
+	require.NoError(t, an1.AddPeer(context.TODO(), *host.InfoFromHost(sn1.host)))
+
+	// sn1 and sn2 subscribe to the topic with the topic validator
+	sub1, err := sn1.Subscribe(topic, authorizedSenderValidator)
+	require.NoError(t, err)
+	sub2, err := sn2.Subscribe(topic)
+	require.NoError(t, err)
+	sub3, err := an1.Subscribe(topic)
+	require.NoError(t, err)
+
+	// assert that the nodes are connected as expected
+	require.Eventually(t, func() bool {
+		return len(sn1.pubSub.ListPeers(topic.String())) > 0 &&
+			len(sn2.pubSub.ListPeers(topic.String())) > 0 &&
+			len(an1.pubSub.ListPeers(topic.String())) > 0
+	}, 3*time.Second, 100*time.Millisecond)
+
+	timedCtx, cancel5s := context.WithTimeout(context.Background(), 1000*time.Second)
+	defer cancel5s()
+	// create a dummy block proposal to publish from our SN node
+	header := unittest.BlockHeaderFixture()
+	data1 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+
+	// sn2 publishes the block proposal, sn1 and an1 should receive the message because
+	// SN nodes are authorized to send block proposals
+	err = sn2.Publish(timedCtx, topic, data1)
+	require.NoError(t, err)
+
+	// sn1 gets the message
+	checkReceive(timedCtx, t, data1, sub1, nil, true)
+
+	// sn2 also gets the message (as part of the libp2p loopback of published topic messages)
+	checkReceive(timedCtx, t, data1, sub2, nil, true)
+
+	// an1 also gets the message
+	checkReceive(timedCtx, t, data1, sub3, nil, true)
+
+	var wg sync.WaitGroup
+	// "eject" sn2 to ensure messages published by ejected nodes get rejected
+	identity2.Ejected = true
+	header = unittest.BlockHeaderFixture()
+	data3 := getMsgFixtureBz(t, &messages.BlockProposal{Header: &header})
+	timedCtx, cancel2s := context.WithTimeout(context.Background(), time.Second)
+	defer cancel2s()
+	err = sn2.Publish(timedCtx, topic, data3)
+	require.NoError(t, err)
+
+	// sn1 should not receive rejected message from ejected sn2
+	timedCtx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	checkReceive(timedCtx, t, nil, sub1, &wg, false)
+
+	unittest.RequireReturnsBefore(t, wg.Wait, 5*time.Second, "could not receive message on time")
+
+	// expecting 1 warn calls for each rejected message from ejected node
+	require.Equalf(t, 1, hookCalls, "expected 1 warning to be logged")
+}
+
+// TestAuthorizedSenderValidator_InvalidWeight tests that the authorized sender validator rejects messages from nodes with invalid weight
+func TestAuthorizedSenderValidator_InvalidWeight(t *testing.T) {
```

Comment:
Mentioned previously, but nodes with 0 weight must be able to participate in limited network communication. We'll need to update this test case.

Reply:
https://github.com/onflow/flow-go/pull/2317/commits/079042c5ef705456cae43752bd375f29afdec63c